graphql_compiler/interpreter/typedefs.py

```diff
@@ -0,0 +1,208 @@
+    - filter_hints: information about the filters applied within the scope relevant to the called
+      function, such as "which filtering operation is being performed?" and "with which arguments?"
+    - neighbor_hints: information about the edges originating from the scope relevant to the called
+      function that the query will eventually need to expand.
+
+    More details on these hints, and suggestions for their use, can be found in the API methods
+    docstrings available below.
+    """
+
+    @abstractmethod
+    def get_tokens_of_type(
+        self,
+        type_name: str,
+        *,
+        runtime_arg_hints: Optional[Mapping[str, Any]] = None,
+        used_property_hints: Optional[AbstractSet[str]] = None,
+        filter_hints: Optional[Collection[FilterInfo]] = None,
+        neighbor_hints: Optional[Collection[Tuple[EdgeInfo, NeighborHint]]] = None,
+        **hints: Any,
+    ) -> Iterable[DataToken]:
+        """Produce an iterable of tokens for the specified type name."""
+        # TODO(predrag): Add more docs in an upcoming PR.
+
+    @abstractmethod
+    def project_property(
+        self,
+        data_contexts: Iterable[DataContext[DataToken]],
+        current_type_name: str,
+        field_name: str,
+        *,
+        runtime_arg_hints: Optional[Mapping[str, Any]] = None,
+        used_property_hints: Optional[AbstractSet[str]] = None,
+        filter_hints: Optional[Collection[FilterInfo]] = None,
+        neighbor_hints: Optional[Collection[Tuple[EdgeInfo, NeighborHint]]] = None,
+        **hints: Any,
+    ) -> Iterable[Tuple[DataContext[DataToken], Any]]:
+        """Produce the values for a given property for each of an iterable of input DataTokens."""
+        # TODO(predrag): Add more docs in an upcoming PR.
+
+    @abstractmethod
+    def project_neighbors(
+        self,
+        data_contexts: Iterable[DataContext[DataToken]],
+        current_type_name: str,
+        edge_info: EdgeInfo,
+        *,
+        runtime_arg_hints: Optional[Mapping[str, Any]] = None,
+        used_property_hints: Optional[AbstractSet[str]] = None,
+        filter_hints: Optional[Collection[FilterInfo]] = None,
+        neighbor_hints: Optional[Collection[Tuple[EdgeInfo, NeighborHint]]] = None,
+        **hints: Any,
+    ) -> Iterable[Tuple[DataContext[DataToken], Iterable[DataToken]]]:
+        """Produce the neighbors along a given edge for each of an iterable of input DataTokens."""
+        # TODO(predrag): Add more docs in an upcoming PR.
+        #
+        # If using a generator instead of a list for the Iterable[DataToken] part,
+        # be careful -- generators are not closures! Make sure any state you pull into
+        # the generator from the outside does not change, or that bug will be hard to find.
+        # Remember: it's always safer to use a function to produce the generator, since
+        # that will explicitly preserve all the external values passed into it.
```

Comment:
```suggestion
        # If using a generator or a mutable data type for the Iterable[DataToken] part,
        # be careful! Make sure any state it depends upon
        # does not change, or that bug will be hard to find.
```
I think this admonition is important enough to warrant keeping it short and sweet.  I don't know that we necessarily need to make additional points about closures and functions, but if we do, I think they might be made more profitably in short code examples than in prose.